components/_impl/workers/in_process_worker.py

```diff
@@ -0,0 +1,49 @@
+import marshal
+import textwrap
+import typing
+
+from components._impl.workers import base
+
+
+class InProcessWorker(base.WorkerBase):
+    """Worker which reuses the current Python process.
+
+    The implementation of this worker borrows from the builtin `timeit.Timer`
+    class, and simply reuses the current interpreter. (Making it comparatively
+    simple.) Note that as a result, it offers no protection against the GIL.
+    """
+
+    def __init__(self, globals: typing.Dict[str, typing.Any]):
+        super().__init__()
+        self._globals: typing.Dict[str, typing.Any] = globals
+
+    @property
+    def in_process(self) -> bool:
+        return True
+
+    def run(self, snippet: str) -> None:
```

Comment:
I guess the idea is that it would look something like: (?)

```
worker = Worker()

@worker.dispatch
def f(x: int, y: int) -> int:
  return x + y
  
z = f(x=1, y=2)
```

That's kind of what `run_in_worker` does now. It's not clear to me how you would handle state between calls. (Could a later dispatch access `x` and `y`?) And I'm worried about both the cognitive and runtime overhead of such a scheme. Maybe something like: (?)
```
def f_store(x: int) -> None:
  pass

def add() -> None:
  # y cannot cross the worker boundary
  y = torch.ones((1,)) + x
  
def f_load() -> int:
  return int(y)

worker = Worker()
worker.run(f_store, x=1)
worker.run(add)
y = worker.run(f_load)
```

Part of the reason for the strong `store`, `run`, `load` separation is that in most cases we expect that we're dealing with objects that we can't pull back into the parent process. So the majority of calls are just manipulating the env, and then occasionally we grab a result. Two things do strike me though, as I think through this again:
1) Maybe `run_in_worker` should be `worker.dispatch`. I feel like I may have already made what you're asking for, but I'm exposing it at the wrong level. (I still think there are cases where the simpler primitives are better just for comprehensibility.)

2) Maybe `store` and `load` should be implemented with `@property` and `.setter`. Instead of:
```
worker.store("x", 1)
worker.run("y = x + 1")
worker.load("y")
```
it's much more Pythonic to write something like:
```
worker.__globals__.x = 1
worker.run("y = x + 1")
y = worker.__globals__.y
```
The key detail that the latter obscures though is that there's a copy, so:
```
worker.__globals__.x = {}
worker.__globals__.x["foo"] = "bar"  # This runs, but does not do what you intend.
```

To wrap up this kind of rambling defense of my beautiful API: it's as pared down as I think it can be before it starts to become confusing (by papering over details of the caller/worker divide), but definitely open to discussion.

Reply:
After some offline discussion, we decided to leave as-is.